hibernate-core/src/main/java/org/hibernate/internal/util/collections/InstanceIdentityMap.java

```
@@ -0,0 +1,377 @@
+	 * since we need to do a type check. Prefer using {@link #containsKey(int, Object)}.
+	 */
+	@Override
+	public boolean containsKey(Object key) {
+		if ( key instanceof InstanceIdentity instance ) {
+			return containsKey( instance.$$_hibernate_getInstanceId(), instance );
+		}
+		throw new ClassCastException( "Provided key does not support instance identity" );
+	}
+
+	@Override
+	public boolean containsValue(Object value) {
+		for ( V v : values() ) {
+			if ( Objects.equals( value, v ) ) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Tests if the specified key-value mapping is in the map.
+	 *
+	 * @param key possible key, must be an instance of {@link InstanceIdentity}
+	 * @param value possible value
+	 * @return {@code true} if and only if the specified key-value mapping is in the map
+	 */
+	private boolean containsMapping(Object key, Object value) {
+		if ( key instanceof InstanceIdentity instance ) {
+			return get( instance.$$_hibernate_getInstanceId(), instance ) == value;
+		}
+		throw new ClassCastException( "Provided key does not support instance identity" );
+	}
+
+	/**
+	 * Returns the value to which the specified instance id is mapped, or {@code null} if this map
+	 * contains no mapping for the instance id.
+	 *
+	 * @param instanceId the instance id whose associated value is to be returned
+	 * @param key key instance to double-check instance equality
+	 * @return the value to which the specified instance id is mapped,
+	 * or {@code null} if this map contains no mapping for the instance id
+	 * @implNote This method accesses the backing array with the provided instance id, but performs an instance
+	 * equality check ({@code ==}) with the provided key to ensure it corresponds to the mapped one
+	 */
+	public @Nullable V get(int instanceId, Object key) {
+		final Entry<K, V> entry = get( instanceId );
+		if ( entry != null && entry.getKey() == key ) {
```

Maybe throw an exception when the keys don't match?